Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,667 @@
+like it currently does through the `initialize()` function of `LambdaRunner` which wraps the handler's `init(...)` and
+handles any errors thrown by reporting it to the dedicated AWS endpoint.
+
+To retain support for initialization error reporting, the `Lambda.reportStartupError(any Error)` function gives users
+the option to manually report initialization errors in their closure handler. Although this should ideally happen
+implicitly like it currently does in v1, we believe this is a small compromise in comparison to the benefits gained in
+now being able to cleanly manage the lifecycles of required services in a structured concurrency manner.
+
+> Use-case:
+>
+> Assume we want to load a secret for the Lambda function from a secret vault first. If this fails, we want to report
+> the error to the control plane:
+>
+> ```swift
+> let secretVault = SecretVault()
+>
+> do {
+>    /// !!! Error thrown: secret "foo" does not exist !!!
+>    let secret = try await secretVault.getSecret("foo")
+>
+>    let runtime = LambdaRuntime { (event: Input, context: LambdaContext) in
+>        /// Lambda business logic
+>    }
+>
+>    let serviceGroup = ServiceGroup(
+>        services: [postgresClient, runtime],
+>        configuration: .init(gracefulShutdownSignals: [.sigterm]),
+>        logger: logger
+>    )
+>    try await serviceGroup.run()
+> } catch {
+>    /// Report startup error straight away to the dedicated initialization error endpoint
+>    try await Lambda.reportStartupError(error)
+> }
+> ```
+
+### Codable support
+
+For v2 we want to introduce a `LambdaCodableAdapter` that:
+
+1. Accepts any generic underlying handler conforming to a new protocol `LambdaHandler` specifying `associatedtype`s
+   `Event` and `Output`, and a `handle(_ event: Event, context: LambdaContext) async throws -> Output` function.
+
+#### LambdaHandler
+
+```swift
+public protocol LambdaHandler {
+    /// Generic input type
+    /// The body of the request sent to Lambda will be decoded into this type for the handler to consume
+    associatedtype Event
+    /// Generic output type
+    /// This is the return type of the handle() function.
+    associatedtype Output
+
+    /// The business logic of the Lambda function. Receives a generic input type and returns a generic output type.
+    /// Agnostic to JSON encoding/decoding
+    mutating func handle(_ event: Event, context: LambdaContext) async throws -> Output
+}
+```
+
+2. Accepts _any_ encoder and decoder object conforming to the `LambdaEventDecoder` and `LambdaOutputEncoder` protocols:
+
+#### LambdaEventDecoder and LambdaOutputEncoder protocols
+
+```swift
+public protocol LambdaEventDecoder {
+    /// Decode the ByteBuffer representing the received event into the generic type T
+    /// the handler will receive
+    func decode<T: Decodable>(_ type: T.Type, from buffer: ByteBuffer) throws -> T
```

Comment:
can't we give type variables nice names like normal variables? `T` could maybe be `Event`?

Reply:
agree! Copy and paste from old code. Should have fixed. @aryan-25 can you fix this please?